In [13]:
import os
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split


In [14]:
flight = pd.read_csv(r'C:\Users\acer\Desktop\Flight-Project\data\flight_price.csv')
flight.head()

Airline Date_of_Journey    Source Destination                  Route  \
0       IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1    Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways       9/06/2019     Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      12/05/2019   Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      01/03/2019  Banglore   New Delhi        BLR → NAG → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info  Price  
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897  
1    05:50         13:15   7h 25m     2 stops         No info   7662  
2    09:25  04:25 10 Jun      19h     2 stops         No info  13882  
3    18:05         23:30   5h 25m      1 stop         No info   6218  
4    16:50         21:35   4h 45m      1 stop         No info  13302

In [15]:
flight.shape

(10683, 11)

In [16]:
flight.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10683 entries, 0 to 10682
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Airline          10683 non-null  object
 1   Date_of_Journey  10683 non-null  object
 2   Source           10683 non-null  object
 3   Destination      10683 non-null  object
 4   Route            10682 non-null  object
 5   Dep_Time         10683 non-null  object
 6   Arrival_Time     10683 non-null  object
 7   Duration         10683 non-null  object
 8   Total_Stops      10682 non-null  object
 9   Additional_Info  10683 non-null  object
 10  Price            10683 non-null  int64 
dtypes: int64(1), object(10)
memory usage: 918.2+ KB


In [17]:
flight.isnull().sum()

Airline            0
Date_of_Journey    0
Source             0
Destination        0
Route              1
Dep_Time           0
Arrival_Time       0
Duration           0
Total_Stops        1
Additional_Info    0
Price              0
dtype: int64

## preliminary analysisi

### 3.1 Check Data Types

In [18]:
flight.head()

Airline Date_of_Journey    Source Destination                  Route  \
0       IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1    Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways       9/06/2019     Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      12/05/2019   Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      01/03/2019  Banglore   New Delhi        BLR → NAG → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info  Price  
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897  
1    05:50         13:15   7h 25m     2 stops         No info   7662  
2    09:25  04:25 10 Jun      19h     2 stops         No info  13882  
3    18:05         23:30   5h 25m      1 stop         No info   6218  
4    16:50         21:35   4h 45m      1 stop         No info  13302

In [19]:
flight.dtypes

Airline            object
Date_of_Journey    object
Source             object
Destination        object
Route              object
Dep_Time           object
Arrival_Time       object
Duration           object
Total_Stops        object
Additional_Info    object
Price               int64
dtype: object

### 3.2 Check for dublicates

In [20]:
flight.duplicated().sum()

220

In [23]:
(
    flight
    .loc[flight.duplicated(keep=False)]
    .sort_values(["Airline","Date_of_Journey","Source","Destination"])
)

Airline Date_of_Journey    Source Destination                  Route  \
6321  Air India      01/03/2019  Banglore   New Delhi  BLR → BOM → AMD → DEL   
9848  Air India      01/03/2019  Banglore   New Delhi  BLR → BOM → AMD → DEL   
572   Air India      03/03/2019  Banglore   New Delhi              BLR → DEL   
8168  Air India      03/03/2019  Banglore   New Delhi              BLR → DEL   
1495  Air India       1/04/2019   Kolkata    Banglore  CCU → DEL → COK → BLR   
...         ...             ...       ...         ...                    ...   
2692   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
2870   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
3711   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
2634    Vistara      24/03/2019  Banglore   New Delhi              BLR → DEL   
5470    Vistara      24/03/2019  Banglore   New Delhi              BLR → DEL   

     Dep_Time  Arrival_Time Duration Total_Stops  \
6321    08:50  23:55 02 Mar   39h 5m     2 stops   
9848    08:50  23:55 02 Mar   39h 5m     2 stops   
572     21:10         23:55   2h 45m    non-stop   
8168    21:10         23:55   2h 45m    non-stop   
1495    10:00  01:20 02 Apr  15h 20m     2 stops   
...       ...           ...      ...         ...   
2692    05:45         08:35   2h 50m    non-stop   
2870    05:45         08:35   2h 50m    non-stop   
3711    20:30         23:20   2h 50m    non-stop   
2634    11:30         14:10   2h 40m    non-stop   
5470    11:30         14:10   2h 40m    non-stop   

                   Additional_Info  Price  
6321                       No info  17135  
9848                       No info  17135  
572                        No info   7591  
8168                       No info   7591  
1495                       No info  10408  
...                            ...    ...  
2692  No check-in baggage included   4273  
2870  No check-in baggage included   4273  
3711  No check-in baggage included   3873  
2634                       No info   5403  
5470                       No info   5403  

[415 rows x 11 columns]

### 3.2 Observations

- object means type is string
- The type of `Date_of_Journey`,`Dep_Time` and `Arrival_Time` is string.It should be datetime
- the type of `Duration` and `Total_Stops` is mixed.It should be numeric
- There are 220 Dublicates.These should be removed

# 4.Detailed Analysis

### Airline

In [37]:
flight["Airline"]

0             IndiGo
1          Air India
2        Jet Airways
3             IndiGo
4             IndiGo
            ...     
10678       Air Asia
10679      Air India
10680    Jet Airways
10681        Vistara
10682      Air India
Name: Airline, Length: 10683, dtype: object

In [38]:
flight.Airline.unique()

array(['IndiGo', 'Air India', 'Jet Airways', 'SpiceJet',
       'Multiple carriers', 'GoAir', 'Vistara', 'Air Asia',
       'Vistara Premium economy', 'Jet Airways Business',
       'Multiple carriers Premium economy', 'Trujet'], dtype=object)

In [44]:
(
    flight
    .Airline
    .str.replace(" Premium economy","")
    .str.replace(" Business","")
    .str.title()
    .unique()
    
)

array(['Indigo', 'Air India', 'Jet Airways', 'Spicejet',
       'Multiple Carriers', 'Goair', 'Vistara', 'Air Asia', 'Trujet'],
      dtype=object)

### Date_of_Journey

In [53]:
flight.Date_of_Journey

0        24/03/2019
1         1/05/2019
2         9/06/2019
3        12/05/2019
4        01/03/2019
            ...    
10678     9/04/2019
10679    27/04/2019
10680    27/04/2019
10681    01/03/2019
10682     9/05/2019
Name: Date_of_Journey, Length: 10683, dtype: object

In [59]:
pd.to_datetime(flight.Date_of_Journey,dayfirst=True)

0       2019-03-24
1       2019-05-01
2       2019-06-09
3       2019-05-12
4       2019-03-01
           ...    
10678   2019-04-09
10679   2019-04-27
10680   2019-04-27
10681   2019-03-01
10682   2019-05-09
Name: Date_of_Journey, Length: 10683, dtype: datetime64[ns]

### Departure time

In [85]:
flight.Dep_Time	

0        22:20
1        05:50
2        09:25
3        18:05
4        16:50
         ...  
10678    19:55
10679    20:45
10680    08:20
10681    11:30
10682    10:55
Name: Dep_Time, Length: 10683, dtype: object

In [87]:
pd.to_datetime(flight.Dep_Time).dt.time

C:\Users\acer\AppData\Local\Temp\ipykernel_8536\2899660464.py:1: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  pd.to_datetime(flight.Dep_Time).dt.time


0        22:20:00
1        05:50:00
2        09:25:00
3        18:05:00
4        16:50:00
           ...   
10678    19:55:00
10679    20:45:00
10680    08:20:00
10681    11:30:00
10682    10:55:00
Name: Dep_Time, Length: 10683, dtype: object

### Duration

In [98]:
flight.Duration

0        2h 50m
1        7h 25m
2           19h
3        5h 25m
4        4h 45m
          ...  
10678    2h 30m
10679    2h 35m
10680        3h
10681    2h 40m
10682    8h 20m
Name: Duration, Length: 10683, dtype: object

In [101]:
(
    flight
    .Duration
    .loc[lambda ser:~ser.str.contains("h")]
)

6474    5m
Name: Duration, dtype: object

In [102]:
flight.iloc[[6474]]

Airline Date_of_Journey  Source Destination                  Route  \
6474  Air India       6/03/2019  Mumbai   Hyderabad  BOM → GOI → PNQ → HYD   

     Dep_Time Arrival_Time Duration Total_Stops Additional_Info  Price  
6474    16:50        16:55       5m     2 stops         No info  17327

In [120]:
(
    flight
    .Duration
    .drop(index=6474)
    .str.split(" ",expand=True)
    .set_axis(["hour","minute"],axis=1)
    .assign(
        hour=lambda df_: (
            df_
            .hour
            .str.replace("h","")
            .astype(int)
            .mul(60)
        ),
        minute=lambda df_: (
            df_
            .minute
            .str.replace("m","")
            .fillna("0")
            .astype(int)
        )
    )
    .sum(axis=1)
    
)

0         170
1         445
2        1140
3         325
4         285
         ... 
10678     150
10679     155
10680     180
10681     160
10682     500
Length: 10682, dtype: int64

In [123]:
(
    flight
    .Duration
    .drop(index=6474)
    .str.split(" ",expand=True)
    .set_axis(["hour","minute"],axis=1)
    .assign(
        hour=lambda df_: (
            df_
            .hour
            .str.replace("h","")
            .astype(int)
            .mul(60)
        ),
        minute=lambda df_: (
            df_
            .minute
            .str.replace("m","")
            .fillna("0")
            .astype(int)
        )
    )
    .sum(axis=1)
    .rename("duration_minutes")
    .to_frame()
    
)

duration_minutes
0                   170
1                   445
2                  1140
3                   325
4                   285
...                 ...
10678               150
10679               155
10680               180
10681               160
10682               500

[10682 rows x 1 columns]

### total stop

In [132]:
flight.Total_Stops.unique()

array(['non-stop', '2 stops', '1 stop', '3 stops', nan, '4 stops'],
      dtype=object)

In [140]:
(
    flight
    .Total_Stops
    .replace("non-stop","0")
    .str.replace(" stops?","",regex=True)
    
    .pipe(lambda ser:pd.to_numeric(ser))
)

0        0.0
1        2.0
2        2.0
3        1.0
4        1.0
        ... 
10678    0.0
10679    0.0
10680    0.0
10681    0.0
10682    2.0
Name: Total_Stops, Length: 10683, dtype: float64

# 5.Cleaning Data

In [126]:
def convert_to_minutes(ser):
    return(
        ser
        .str.split(" ",expand=True)
        .set_axis(["hour","minute"],axis=1)
        .assign(
            hour=lambda df_: (
                df_
                .hour
                .str.replace("h","")
                .astype(int)
                .mul(60)
            ),
            minute=lambda df_: (
                df_
                .minute
                .str.replace("m","")
                .fillna("0")
                .astype(int)
            )
        )
        .sum(axis=1)
        
    )

In [146]:
def clean_data(df):

    return (
        df
         .drop(index=[6474])
         .assign(**{
             col:df[col].str.strip()
             for col in df.select_dtypes(include="O").columns
         })
        .rename(columns=str.lower)
        .assign(
            airline=lambda df_: (
                
                    df_
                   .airline
                   .str.replace(" Premium economy","")
                   .str.replace(" Business","")
                   .str.title()
                   
    
                ),
                date_of_journey=lambda df_: pd.to_datetime(df_.date_of_journey, dayfirst=True),
                dep_time=lambda df_: pd.to_datetime(df_.dep_time).dt.time,
                arrival_time=lambda df_: pd.to_datetime(df_.arrival_time).dt.time,
                duration=lambda df_:df_.duration.pipe(convert_to_minutes),
                total_stops=lambda df_:(
                    df_
                    .total_stops
                     .replace("non-stop","0")
                     .str.replace(" stops?","",regex=True)
    
                     .pipe(lambda ser:pd.to_numeric(ser))
                    
                ),
                additional_info=lambda df_:df_.additional_info.replace("No info","No Info")
                   
                  
                
             )
          
          .drop(columns="route")
    
    )

In [147]:
print(flight.columns)


Index(['Airline', 'Date_of_Journey', 'Source', 'Destination', 'Route',
       'Dep_Time', 'Arrival_Time', 'Duration', 'Total_Stops',
       'Additional_Info', 'Price'],
      dtype='object')


In [150]:
flight_cleaned=clean_data(flight)
flight_cleaned

C:\Users\acer\AppData\Local\Temp\ipykernel_8536\829455626.py:23: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  dep_time=lambda df_: pd.to_datetime(df_.dep_time).dt.time,
C:\Users\acer\AppData\Local\Temp\ipykernel_8536\829455626.py:24: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  arrival_time=lambda df_: pd.to_datetime(df_.arrival_time).dt.time,


airline date_of_journey    source destination  dep_time  \
0           Indigo      2019-03-24  Banglore   New Delhi  22:20:00   
1        Air India      2019-05-01   Kolkata    Banglore  05:50:00   
2      Jet Airways      2019-06-09     Delhi      Cochin  09:25:00   
3           Indigo      2019-05-12   Kolkata    Banglore  18:05:00   
4           Indigo      2019-03-01  Banglore   New Delhi  16:50:00   
...            ...             ...       ...         ...       ...   
10678     Air Asia      2019-04-09   Kolkata    Banglore  19:55:00   
10679    Air India      2019-04-27   Kolkata    Banglore  20:45:00   
10680  Jet Airways      2019-04-27  Banglore       Delhi  08:20:00   
10681      Vistara      2019-03-01  Banglore   New Delhi  11:30:00   
10682    Air India      2019-05-09     Delhi      Cochin  10:55:00   

      arrival_time  duration  total_stops additional_info  price  
0         01:10:00       170          0.0         No Info   3897  
1         13:15:00       445          2.0         No Info   7662  
2         04:25:00      1140          2.0         No Info  13882  
3         23:30:00       325          1.0         No Info   6218  
4         21:35:00       285          1.0         No Info  13302  
...            ...       ...          ...             ...    ...  
10678     22:25:00       150          0.0         No Info   4107  
10679     23:20:00       155          0.0         No Info   4145  
10680     11:20:00       180          0.0         No Info   7229  
10681     14:10:00       160          0.0         No Info  12648  
10682     19:15:00       500          2.0         No Info  11753  

[10682 rows x 10 columns]

# split the data

In [153]:
flight_final = flight_cleaned.sample(1000)

In [154]:
X = flight_final.drop(columns="price")
y = flight_final.price.copy()

In [155]:
X_, X_test, y_, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
X_train, X_val, y_train, y_val = train_test_split(X_, y_, test_size=0.2, random_state=42)

print(X_train.shape, y_train.shape)
print(X_val.shape, y_val.shape)
print(X_test.shape, y_test.shape)

(640, 9) (640,)
(160, 9) (160,)
(200, 9) (200,)


In [160]:
PROJECT_DIR = "C:\\Users\\acer\\Desktop\\Flight-Project"
DATA_DIR = "data"

def export_data(X, y, name):
	file_name = f"{name}.csv"
	file_path = os.path.join(PROJECT_DIR, DATA_DIR, file_name)

	X.join(y).to_csv(file_path, index=False)

	return pd.read_csv(file_path).head()

In [161]:
export_data(X_train, y_train, "train")

airline date_of_journey    source destination  dep_time arrival_time  \
0      Vistara      2019-04-15  Banglore       Delhi  09:45:00     12:35:00   
1  Jet Airways      2019-03-18  Banglore   New Delhi  11:40:00     07:40:00   
2  Jet Airways      2019-05-03    Mumbai   Hyderabad  10:20:00     11:50:00   
3       Indigo      2019-04-03  Banglore       Delhi  13:00:00     15:50:00   
4  Jet Airways      2019-05-12   Kolkata    Banglore  16:30:00     18:15:00   

   duration  total_stops              additional_info  price  
0       170          0.0                      No Info   5403  
1      1200          1.0                      No Info  12284  
2        90          0.0                      No Info   8040  
3       170          0.0                      No Info   4077  
4      1545          1.0  In-flight meal not included  10844

In [162]:
export_data(X_val, y_val, "val")

airline date_of_journey    source destination  dep_time arrival_time  \
0     Spicejet      2019-04-06   Kolkata    Banglore  22:20:00     00:40:00   
1  Jet Airways      2019-05-01   Kolkata    Banglore  18:55:00     08:15:00   
2  Jet Airways      2019-05-24   Kolkata    Banglore  09:35:00     09:45:00   
3       Indigo      2019-03-01   Kolkata    Banglore  21:25:00     00:05:00   
4      Vistara      2019-05-24  Banglore       Delhi  07:00:00     09:40:00   

   duration  total_stops               additional_info  price  
0       140          0.0  No check-in baggage included   3841  
1       800          1.0   In-flight meal not included   8586  
2      1450          1.0                       No Info  13067  
3       160          0.0                       No Info   4778  
4       160          0.0                       No Info   4668

In [163]:
export_data(X_test, y_test, "test")

airline date_of_journey   source destination  dep_time  \
0  Multiple Carriers      2019-03-06    Delhi      Cochin  13:00:00   
1            Vistara      2019-05-21  Kolkata    Banglore  17:00:00   
2        Jet Airways      2019-05-18  Kolkata    Banglore  05:45:00   
3  Multiple Carriers      2019-03-21    Delhi      Cochin  09:45:00   
4          Air India      2019-05-24  Kolkata    Banglore  09:25:00   

  arrival_time  duration  total_stops              additional_info  price  
0     01:35:00       755          1.0                      No Info  15603  
1     20:20:00      1640          1.0                      No Info   9345  
2     19:50:00       845          1.0  In-flight meal not included  10844  
3     16:10:00       385          1.0                      No Info   7242  
4     13:45:00      1700          2.0                      No Info  12592